In [33]:
import pandas as pd
import numpy as np
import warnings
import pickle
from sklearn.metrics.pairwise import cosine_similarity
warnings.filterwarnings("ignore")

## Reading datasets

In [2]:
df_books = pd.read_csv('C:\\Users\\Priyaranjan\\Desktop\\Kaggle datasets\\Books dataset - Recommender systems\\Books.csv')
df_users = pd.read_csv('C:\\Users\\Priyaranjan\\Desktop\\Kaggle datasets\\Books dataset - Recommender systems\\Users.csv')
df_ratings = pd.read_csv('C:\\Users\\Priyaranjan\\Desktop\\Kaggle datasets\\Books dataset - Recommender systems\\Ratings.csv')

In [3]:
df_books.head(2)

ISBN           Book-Title           Book-Author Year-Of-Publication  \
0  0195153448  Classical Mythology    Mark P. O. Morford                2002   
1  0002005018         Clara Callan  Richard Bruce Wright                2001   

                 Publisher                                        Image-URL-S  \
0  Oxford University Press  http://images.amazon.com/images/P/0195153448.0...   
1    HarperFlamingo Canada  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...

In [4]:
df_users.head(2)

User-ID                   Location   Age
0        1         nyc, new york, usa   NaN
1        2  stockton, california, usa  18.0

In [5]:
df_ratings.head(2)

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5

In [6]:
print(df_books.shape)
print(df_users.shape)
print(df_ratings.shape)

(271360, 8)
(278858, 3)
(1149780, 3)


In [7]:
df_books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

We can see very minimal number of rows having null ins ome columns which we can ignore for now as eventually we would be producing a sparse matrix.

In [8]:
df_users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

Column 'Age' is null for almost 50% of the rows. At this point, we know we wont be using 'Age' in our recommender system. Hence, ignoring the missing values here for now.

In [9]:
df_ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [10]:
print(df_books.duplicated().sum())
print(df_users.duplicated().sum())
print(df_ratings.duplicated().sum())

0
0
0


## Popularity based recommender system

Firstly, merging the books & ratings dataframes.

In [11]:
df_ratings_books = df_ratings.merge(df_books,on='ISBN')
df_ratings_books.shape

(1031136, 10)

In [12]:
df_ratings_books.head(2)

User-ID        ISBN  Book-Rating            Book-Title Book-Author  \
0   276725  034545104X            0  Flesh Tones: A Novel  M. J. Rose   
1     2313  034545104X            5  Flesh Tones: A Novel  M. J. Rose   

  Year-Of-Publication         Publisher  \
0                2002  Ballantine Books   
1                2002  Ballantine Books   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/034545104X.0...   
1  http://images.amazon.com/images/P/034545104X.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/034545104X.0...  
1  http://images.amazon.com/images/P/034545104X.0...

In [13]:
df_num_rating = df_ratings_books.groupby('Book-Title').count()['Book-Rating'].reset_index()
df_num_rating.rename(columns={'Book-Rating':'num_ratings'},inplace=True)
df_num_rating

Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [14]:
df_ratings_books1 = df_ratings_books[['Book-Title','Book-Rating']]
df_ratings_books1.head(2)

Book-Title  Book-Rating
0  Flesh Tones: A Novel            0
1  Flesh Tones: A Novel            5

In [15]:
df_avg_rating = df_ratings_books1.groupby('Book-Title').mean()['Book-Rating'].reset_index()
df_avg_rating.rename(columns={'Book-Rating':'avg_ratings'},inplace=True)
df_avg_rating

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [16]:
df_popular = df_num_rating.merge(df_avg_rating,on='Book-Title')
df_popular.head(2)

Book-Title  num_ratings  avg_ratings
0   A Light in the Storm: The Civil War Diary of ...            4         2.25
1                              Always Have Popsicles            1         0.00

As we need the popular books to be listed here, we will be filtering the above df based on num_ratings.
- Keep only the books which have num_ratings > 200
- This way, we are ensuring that when the recommender suggests teh user with popular books, it will only consider which has more than 200 ratings on them as it shows that the book has been read & rated by huge number of users.
- Lastly we would need the TOP 50 books out of the descendingly sorted values for avg_rating for each book.

In [17]:
df_popular = df_popular[df_popular['num_ratings'] >= 200].sort_values('avg_ratings',ascending=False).head(50)
df_popular

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
60582       Ender's Game (Ender Wiggins Saga (Paperback))          249   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
191612  The Hobbit : The Enchanting Prelude to The Lor...          281   
187377  The Fellowship of the Ring (The Lord of the Ri...          368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211384     The Two Towers (The Lord of the Rings, Part 2)          260   
219741                              To Kill a Mockingbird          510   
183573                                  The Da Vinci Code          898   
187880                 The Five People You Meet in Heaven          430   
180556                             The Catcher in the Rye          449   
204672  The Return of the King (The Lord of the Rings,...          211   
196326                          The Lovely Bones: A Novel         1295   
190081                                   The Great Gatsby          205   
764                                                  1984          284   
144165                           Prodigal Summer: A Novel          253   
128670                                         Neverwhere          265   
206502                            The Secret Life of Bees          774   
168719  Stupid White Men ...and Other Sorry Excuses fo...          283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204387                The Red Tent (Bestselling Backlist)          723   
27295                                     Brave New World          226   
191589               The Hitchhiker's Guide to the Galaxy          268   
129379   Nickel and Dimed: On (Not) Getting By in America          335   
209778                                    The Tao of Pooh          211   
14607                                         Animal Farm          233   
93381                                       Into the Wild          252   
63867                                      Fahrenheit 451          409   
7373    A Walk in the Woods: Rediscovering America on ...          205   
74750                           Girl with a Pearl Earring          526   
129444                                              Night          202   
136145                                          Outlander          283   
209209      The Subtle Knife (His Dark Materials, Book 2)          225   
181699  The Color of Water: A Black Man's Tribute to H...          223   
233370  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156102                     Seabiscuit: An American Legend          275   
107962                                         Life of Pi          664   
176845                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8434                                       ANGELA'S ASHES          279   
83432                                       High Fidelity          247   
76343                                         Good in Bed          490   
64931   Fast Food Nation: The Dark Side of the All-Ame...          321   
12700                                       American Gods          302   
161645                                 Skipping Christmas          322   
105777  Left Behind: A Novel of the Earth's Last Days ...          318   
189551    The Golden Compass (His Dark Materials, Book 1)          336   

        avg_ratings  
80434      5.852804  
80422      5.824289  
80441      5.737410  
80426      5.501441  
60582      5.409639  
80414      5.183453  
191612     5.007117  
187377     4.948370  
80445      4.895652  
211384     4.880769  
219741     4.700000  
183573  

Getting the other attributes of these books from df_books.

In [18]:
df_popular = df_popular.merge(df_books,on='Book-Title').drop_duplicates('Book-Title')[['Book-Title', 
                                                                                       'Book-Author', 
                                                                                       'Image-URL-M', 
                                                                                       'num_ratings', 
                                                                                       'avg_ratings']]
df_popular

Book-Title          Book-Author  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)        J. K. Rowling   
3         Harry Potter and the Goblet of Fire (Book 4)        J. K. Rowling   
5       Harry Potter and the Sorcerer's Stone (Book 1)        J. K. Rowling   
9    Harry Potter and the Order of the Phoenix (Boo...        J. K. Rowling   
13       Ender's Game (Ender Wiggins Saga (Paperback))     Orson Scott Card   
17    Harry Potter and the Chamber of Secrets (Book 2)        J. K. Rowling   
20   The Hobbit : The Enchanting Prelude to The Lor...       J.R.R. TOLKIEN   
21   The Fellowship of the Ring (The Lord of the Ri...       J.R.R. TOLKIEN   
30   Harry Potter and the Sorcerer's Stone (Harry P...        J. K. Rowling   
32      The Two Towers (The Lord of the Rings, Part 2)       J.R.R. TOLKIEN   
43                               To Kill a Mockingbird           Harper Lee   
51                                   The Da Vinci Code            Dan Brown   
57                  The Five People You Meet in Heaven          Mitch Albom   
59                              The Catcher in the Rye        J.D. Salinger   
66   The Return of the King (The Lord of the Rings,...       J.R.R. TOLKIEN   
76                           The Lovely Bones: A Novel         Alice Sebold   
77                                    The Great Gatsby  F. Scott Fitzgerald   
79                                                1984        George Orwell   
88                            Prodigal Summer: A Novel   Barbara Kingsolver   
89                                          Neverwhere          Neil Gaiman   
94                             The Secret Life of Bees        Sue Monk Kidd   
100  Stupid White Men ...and Other Sorry Excuses fo...        Michael Moore   
101  Tuesdays with Morrie: An Old Man, a Young Man,...          MITCH ALBOM   
104                The Red Tent (Bestselling Backlist)        Anita Diamant   
105                                    Brave New World        Aldous Huxley   
110               The Hitchhiker's Guide to the Galaxy        Douglas Adams   
119   Nickel and Dimed: On (Not) Getting By in America   Barbara Ehrenreich   
121                                    The Tao of Pooh        Benjamin Hoff   
123                                        Animal Farm        George Orwell   
133                                      Into the Wild         Jon Krakauer   
136                                     Fahrenheit 451         Ray Bradbury   
149  A Walk in the Woods: Rediscovering America on ...          Bill Bryson   
150                          Girl with a Pearl Earring      Tracy Chevalier   
151                                              Night          Elie Wiesel   
156                                          Outlander       DIANA GABALDON   
162      The Subtle Knife (His Dark Materials, Book 2)       PHILIP PULLMAN   
170  The Color of Water: A Black Man's Tribute to H...        James McBride   
172  Where the Heart Is (Oprah's Book Club (Paperba...         Billie Letts   
173                     Seabiscuit: An American Legend    LAURA HILLENBRAND   
175                                         Life of Pi          Yann Martel   
179                                     The Bean Trees   Barbara Kingsolver   
182  A Child Called \It\": One Child's Courage to S...          Dave Pelzer   
184                                     ANGELA'S ASHES        Frank McCourt   
186                                      High Fidelity          Nick Hornby   
192                                        Good in Bed      Jennifer Weiner   
194  Fast Food Nation: The Dark Side of the All-Ame...       Eric Schlosser   
195                                      American Gods          Neil Gaiman   
196                                 Skipping Christmas         JOHN GRISHAM   
202  Left Behind: A Novel of the Earth's Last Days ...           Tim Lahaye   
204    The Golden Compass (His Dark Materials, Book 1)       PHILIP PULLMAN   

              

In [19]:
df_popular.shape

(50, 5)

## Collaborative Filtering based recommender system

Here we are going ahead with the idea of building our recommendation system intelligent enough to recommend good items/books. 

Hence, we will be only considering the books that have been rated atleast by 50 people & users who have rated atleast 200 times.

In [20]:
filtered_users_temp = df_ratings_books.groupby('User-ID').count()['Book-Rating'] >= 200
filtered_users_index = filtered_users_temp[filtered_users_temp].index

In [21]:
filtered_users_ratings = df_ratings_books[df_ratings_books['User-ID'].isin(filtered_users_index)]
filtered_users_ratings

User-ID        ISBN  Book-Rating  \
2           6543  034545104X            0   
5          23768  034545104X            0   
7          28523  034545104X            0   
15         77940  034545104X            0   
16         81977  034545104X            0   
...          ...         ...          ...   
1030883   275970  1880837927            0   
1030884   275970  188717897X            0   
1030885   275970  1888889047            0   
1030886   275970  1931868123            0   
1030887   275970  3411086211           10   

                                                Book-Title  \
2                                     Flesh Tones: A Novel   
5                                     Flesh Tones: A Novel   
7                                     Flesh Tones: A Novel   
15                                    Flesh Tones: A Novel   
16                                    Flesh Tones: A Novel   
...                                                    ...   
1030883                         The Theology of the Hammer   
1030884  The Ordeal of Integration: Progress and Resent...   
1030885  Pushcart's Complete Rotten Reviews &amp; Rejec...   
1030886  There's a Porcupine in My Outhouse: Misadventu...   
1030887                                         Die Biene.   

                           Book-Author Year-Of-Publication  \
2                           M. J. Rose                2002   
5                           M. J. Rose                2002   
7                           M. J. Rose                2002   
15                          M. J. Rose                2002   
16                          M. J. Rose                2002   
...                                ...                 ...   
1030883                 Millard Fuller                1994   
1030884              Orlando Patterson                1998   
1030885                 Bill Henderson                1998   
1030886                   Mike Tougias                2002   
1030887  Sybil Gr�¤fin Sch�¶nfeldt                1993   

                                    Publisher  \
2                            Ballantine Books   
5                            Ballantine Books   
7                            Ballantine Books   
15                           Ballantine Books   
16                           Ballantine Books   
...                                       ...   
1030883         Smyth &amp; Helwys Publishing   
1030884                Civitas Book Publisher   
1030885                        Pushcart Press   
1030886                    Capital Books (VA)   
1030887  Bibliographisches Institut, Mannheim   

                                               Image-URL-S  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.amazon.com/images/P/1931868123.0...   
1030887  http://images.amazon.com/images/P/3411086211.0...   

                                               Image-URL-M  \
2        http://images.amazon.com/images/P/034545104X.0...   
5        http://images.amazon.com/images/P/034545104X.0...   
7        http://images.amazon.com/images/P/034545104X.0...   
15       http://images.amazon.com/images/P/034545104X.0...   
16       http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1030883  http://images.amazon.com/images/P/1880837927.0...   
1030884  http://images.amazon.com/images/P/188717897X.0...   
1030885  http://images.amazon.com/images/P/1888889047.0...   
1030886  http://images.a

In [22]:
filtered_books = filtered_users_ratings.groupby('Book-Title').count()['Book-Rating']>=50
filtered_books_index = filtered_books[filtered_books].index

In [23]:
final_ratings = filtered_users_ratings[filtered_users_ratings['Book-Title'].isin(filtered_books_index)]

In [24]:
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58823 r

After dropping duplicates, moving ahead with creating the pivot table or user-item matrix which is requried for our recommender.

In [25]:
#final_ratings.drop_duplicates()
pt = final_ratings.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')
pt.fillna(0,inplace = True)

In [26]:
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

Now, lets move ahead with finding similarity scores for our books based on euclidean distances.

In [27]:
similarity_scores = cosine_similarity(pt)
similarity_scores.shape

(707, 707)

Each book is being compared with all the 707 books to find the similarity scores.

Now, lets create our Recommender function to recommend 5 books when provided with a book name as input.

### Recommender function

In [30]:
def recommend_books(book_title):
    #find the index position of book given as input
    idx = np.where(pt.index == book_title)[0][0]
    # get the similarity scores for that index & fetch the indices of top 5 similar books 
    similar_books_idx = sorted(list(enumerate(similarity_scores[idx])),
                              key = lambda x: x[1], reverse = True)[1:6]
    
    # print out the book titles on the similar books indices
    for i in similar_books_idx:
        print(pt.index[i[0]])

In [32]:
recommend_books('Life of Pi')

Tell No One
The Lovely Bones: A Novel
Fried Green Tomatoes at the Whistle Stop Cafe
Fall On Your Knees (Oprah #45)
Balzac and the Little Chinese Seamstress : A Novel


Dumping the pickle files:

In [34]:
pickle.dump(df_popular,open('popular.pkl','wb'))
pickle.dump(pt,open('pt.pkl','wb'))
pickle.dump(df_books,open('books.pkl','wb'))
pickle.dump(similarity_scores,open('similarity_scores.pkl','wb'))